rfcs/0039-snark-keys-management.md

```diff
@@ -0,0 +1,208 @@
+  particular keys that any build is using, as well as preventing us from using
+  a single, unified binary for our different configurations.
+* Every set of keys that have been generated for a CI job are stored in the S3
+  keys bucket, with no easy way to identify them.
+* Keys in the S3 bucket are always preferred to locally-generated keys during
+  the build process.
+  - During local builds from developers' machines, internet connection problems
+    may result in either partially-downloaded keyfiles or long download waits
+    while the keys are fetched from S3.
+* It is difficult for the infrastructure team to find and package the correct
+  keys for a particular binaries when deploying.
+* The caching behavior for keys is implicit and entangled with the pickles
+  proof system interface, which makes it hard for the team to inspect or adjust
+  the behavior.
+
+The goal of this RFC is to resolve these issues. At a high-level, the proposed
+steps to do this are:
+* Make keys identifiable from the file contents.
+* Add a tool for generating keys outside of the build process.
+* Allow new keys to be specified explicitly.
+* Make key generation explicit, removing it from the build process.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+### Make keys identifiable from the file contents
+
+In order to confirm that loaded keys are compatible with a given configuration,
+we need to have some representation of the configuration embedded with the keys
+files. We propose that every key (and URS) file contains a header in minified
+JSON format before the actual contents, specifying
+* the constraint constants, as defined in the runtime configuration
+* the SHA commit identifiers for the current commit in the coda, marlin and
+  zexe repos used when generating the keys
+* the length of the binary data following the header, so that we can validate
+  that the file was successfully downloaded in full, where applicable
+* the date when the file's contents were created
+* the type of the file's contents
+* the constraint system hash (for keys) or domain size (for URS)
+* any other input information
+  - e.g. constraint system hash for the transaction snark wrapped by a
+    particular blockchain snark
+* a version number, so that we can add or modify fields
+* the identifying hash, which should match the hash part of the filename for
+  generated files
+
+For example, a header for the transaction snark key file might look like:
+```json
+{ "header_version": 1
+, "kind":
+    { "type": "step_proving_key"
+    , "identifier": "transaction_snark_base" }
+, "constraint_constants":
+    { "c": 8
+    , "ledger_depth": 14
+    , "work_delay": 2
+    , "block_window_duration_ms": 180000
+    , "transaction_capacity": {"txns_per_second_x10": "2"}
+    , "coinbase_amount": "200"
+    , "supercharged_coinbase_factor": 2
+    , "account_creation_fee": "0.001" }
+, "commits":
+    { "coda": "COMMIT_SHA_HERE"
```

Comment:
*mina

Reply:
Also the marlin/zexe commits should be redundant since they're specified by the submodule checked out at that commit, right?

Reply:
> the marlin/zexe commits should be redundant

I've removed the zexe commit. Keeping the marlin one because we're using it as an input to the 'identifying hash', and it would be nice if we could compute/check that from the header without needing access to the full git history. I've added a sub-bullet with this justification.